crypto/threads_common.c

```diff
@@ -0,0 +1,451 @@
+/*
+ * Copyright 2025 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/**
+ * @file
+ * @brief Thread-local context-specific data management for OpenSSL
+ *
+ * This file implements a mechanism to store and retrieve context-specific
+ * data using OpenSSL's thread-local storage (TLS) system. It provides a way
+ * to associate and manage data based on a combination of a thread-local key
+ * and an `OSSL_LIB_CTX *` context.
+ *
+ * NOTE: This differs from the CRYPTO_THREAD_[get|set]_local api set in that
+ * this api stores a single OS level thread-local key per-thread, and manages
+ * subsequent keys using a series of sparse arrays stored against that aforementioned
+ * thread local key
+ *
+ * Data Design:
+ *
+ * per-thread master key data->  ┌──────────────┬─┐
+ *                               │              │ │
+ *                               │              │ │
+ *                               └──────────────┘ │
+ *                               ┌──────────────┐ │
+ *                               │              │ │
+ *                               │              │ │
+ *                               └──────────────┘ │  sparse  array indexed
+ *                                      .         │      by key id
+ *                                      .         │
+ *                                      .         │
+ *                               ┌──────────────┐ │
+ *                               │              │ │
+ *                               │       │      │ │
+ *                               └───────┼──────┴─┘
+ *                                       │
+ *          ┌┬───────────────┐           │
+ *          ││               │◄──────────┘
+ *          ││               │
+ *          │└───────────────┘
+ *          │┌───────────────┐
+ *          ││               │
+ *          ││               │  sparse array indexed
+ *          │└───────────────┘     by libctx pointer
+ *          │        .              cast to uintptr_t
+ *          │        .
+ *          │        .
+ *          │┌───────────────┐
+ *          ││        ┌──────┼────► ┌─────────────────┐
+ *          ││        │      │      │                 │
+ *          └┴────────┴──────┘      └─────────────────┘
+ *                                  per-<thread*ctx> data
+ * It uses sparse arrays to map:
+ *   - Thread-local key IDs to master key entries.
+ *   - Library context pointers to context-specific data.
+ *
+ * Two primary functions are provided:
+ *   - CRYPTO_THREAD_get_local_ex() retrieves data associated with a key and
+ *     context.
+ *   - CRYPTO_THREAD_set_local_ex() associates data with a given key and
+ *     context, allocating tables as needed.
```

Comment:
I think we will need to add explicit _create() and _destroy() functions. This will allow function _set() to fail if we attempt to overwrite value at existing key.